unit/unit_fmq/unit_fmq.c

```diff
@@ -0,0 +1,586 @@
+}
+
+/*==========================================================================*
+ * read/write counters
+ *==========================================================================*/
+
+static
+void
+test_read_write_counters(
+    void)
+{
+    gint max_num_items = 8;
+    gint write_count = 6;
+    gint64 in_data[max_num_items];
+    gint64 out_data[max_num_items];
+    guint i;
+    GBinderFmq* fmq = gbinder_fmq_new(sizeof(gint64), max_num_items,
+        GBINDER_FMQ_TYPE_SYNC_READ_WRITE, GBINDER_FMQ_FLAG_CONFIGURE_EVENT_FLAG,
+        -1, 0);
+
+    /* Intiailize input data with random numbers */
+    for (i = 0; i < max_num_items; ++i) {
+        in_data[i] = g_random_int();
+    }
+    memset(out_data, 0, max_num_items);
+
+    /* Write data one value at a time */
+    for (i = 0; i < write_count; ++i) {
+        g_assert(gbinder_fmq_write(fmq, &in_data[i], 1));
+        g_assert(gbinder_fmq_available_to_read(fmq) == i + 1);
+        g_assert(gbinder_fmq_available_to_write(fmq) ==
+            max_num_items - i - 1);
+    }
+
+    /* Read data one value at a time */
+    for (i = 0; i < 2; ++i) {
+        g_assert(gbinder_fmq_read(fmq, &out_data, 1));
+        g_assert(out_data[0] == in_data[i]);
+        g_assert(gbinder_fmq_available_to_read(fmq) ==
+            write_count - i - 1);
+        g_assert(gbinder_fmq_available_to_write(fmq) ==
+            max_num_items - write_count + i + 1);
+        g_assert(gbinder_fmq_available_to_write_contiguous(fmq) ==
+            max_num_items - write_count);
+    }
+
+    g_assert(gbinder_fmq_read(fmq, &out_data, 2));
+    g_assert(gbinder_fmq_available_to_read(fmq) == 2);
+    g_assert(gbinder_fmq_available_to_write(fmq) == 6);
+    g_assert(gbinder_fmq_available_to_write_contiguous(fmq) == 2);
+    g_assert(gbinder_fmq_write(fmq, in_data, 4));
+    g_assert(gbinder_fmq_available_to_read(fmq) == 6);
+    g_assert(gbinder_fmq_available_to_read_contiguous(fmq) == 4);
+    g_assert(gbinder_fmq_read(fmq, &out_data, 6));
+    g_assert(gbinder_fmq_available_to_read(fmq) == 0);
+    g_assert(gbinder_fmq_available_to_read_contiguous(fmq) == 0);
+    g_assert(gbinder_fmq_write(fmq, in_data, 6));
+    g_assert(gbinder_fmq_available_to_read(fmq) == 6);
+    g_assert(gbinder_fmq_available_to_read_contiguous(fmq) == 6);
+    g_assert(gbinder_fmq_available_to_write(fmq) == 2);
+    g_assert(gbinder_fmq_available_to_write_contiguous(fmq) == 2);
+
+    gbinder_fmq_unref(fmq);
+}
+
+/*==========================================================================*
+ * read/write external fd
+ *==========================================================================*/
+
+static
+void
+test_read_write_external_fd(
+    void)
+{
+    gint max_num_items = 8;
+    gint64 in_data[max_num_items];
+    gint64 out_data[max_num_items];
+    gsize item_size = sizeof(gint64);
+    guint i;
+    GBinderFmq* fmq;
+
+    /* Allocate shared memory */
+
+    int shmem_fd = memfd_create("MessageQueueData", MFD_CLOEXEC);
```

Comment:
Same `MFD_CLOEXEC`/`memfd_create` issue that I complained about in `src/gbinder_fmq.c` except that here it would also require `#include <sys/syscall.h>`